pkg/protodep/protodep.go

```diff
@@ -0,0 +1,46 @@
+package protodep
+
+import (
+	"context"
+)
+
+//go:generate bash ./generate.sh
+
+const (
+	DefaultDepDir = "vendor"
+)
+
+type DepFactory interface {
+	Ensure(ctx context.Context, opts *Config) error
+}
+
+type Manager struct {
+	depFactories []DepFactory
+}
+
+func NewManager(ctx context.Context, cwd string) (*Manager, error) {
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	goMod, err := NewGoModFactory(cwd)
+	if err != nil {
+		return nil, err
+	}
+	return &Manager{
+		depFactories: []DepFactory{
+			goMod,
+		},
+	}, nil
+}
+
+func (m *Manager) Ensure(ctx context.Context, opts *Config) error {
+	if err := opts.Validate(); err != nil {
+		return err
+	}
+	for _, v := range m.depFactories {
+		if err := v.Ensure(ctx, opts); err != nil {
+			return err
+		}
+	}
+	return nil
+}
```

Comment:
since `protodep` package seems to be the core of this PR, i think it should be split into a separate PR and commented adequately so it can be reviewed properly. the scope of the PR in its current form is too large to evaluate the impact of all these changes at once